File: SEM 6/Computer Networks/data_structure/stack.lua
Stack = {}


function Stack.is_empty(self)
    return #self.data == 0
end


-- Add an element to the stack
function Stack.push(self, value)
    self.data[#self.data + 1] = value
end


-- Remove and return the top element
function Stack.pop(self)
    if self:is_empty() then
        return nil, "stack is empty"
    end
    return table.remove(self.data)
end

-- Get the top element without removing it
function Stack.top(self)
    if self:is_empty() then
        return nil, "stack is empty"
    end
    return self.data[#self.data]
end

-- Display the current state of stack
function Stack.display(self)
    for i = 1, #self.data do 
        print(self.data[i])
    end
end

function Stack.init()
    local instance = {data = {}} -- Create a new table with `data`
    setmetatable(instance, {__index = Stack}) -- Set Stack as the metatable
    return instance
end




-- ############

local stack = Stack:init()
stack:push(10)
stack:push(20)
-- print(stack:pop())
-- print(stack:top())
stack:display()

File: SEM 6/Computer Networks/data_structure/stack.py
class Stack:
    def __init__(self):
        self.stack = []
    
    # Insert an element
    def push(self, ele):
        self.stack.append(ele)
        
    # Removing an element.
    def pop(self):
        # if self.is_empty():
        #     return "Stack is empty"
        assert(not self.is_empty())
        self.stack.pop()
        
    
    # Shows the top element.
    def top(self):
        # if self.is_empty():
        #     return "Stack is empty"
        assert(not self.is_empty())
        return self.stack[-1]
    
    # Check Stack is empty or not.
    def is_empty(self):
        return len(self.stack) == 0
       
    
    # Display the current state of stack.
    def display(self):
        for i in range(len(self.stack)):
            print(self.stack[i], "--> ", end="")
        print()
            

#####################################
stack = Stack()
stack.push(1)
stack.push(2)
stack.display()
stack.pop()
stack.display()
stack.pop()

